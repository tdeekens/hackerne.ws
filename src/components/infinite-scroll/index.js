export { default } from './infinite-scroll';
