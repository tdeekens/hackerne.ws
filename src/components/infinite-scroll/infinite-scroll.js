import React from 'react';
import PropTypes from 'prop-types';

export class InfiniteScroll extends React.Component {
  static displayName = 'InfiniteScroll';
  static propTypes = {
    isLoading: PropTypes.bool,
    onFetchMore: PropTypes.func.isRequired,
    sentinelRef: PropTypes.shape({ current: PropTypes.object }).isRequired,
    children: PropTypes.node.isRequired,
  };
  static defaultProps = { isLoading: false };

  state = {};

  handleObserver = (entities, observer) => {
    const y = entities[0].boundingClientRect.y;

    if (this.state.prevY > y) {
      if (!this.props.isLoading) {
        this.props.onFetchMore();
      }
    }
    this.setState({ prevY: y });
  };

  componentDidMount() {
    this.observer = new IntersectionObserver(this.handleObserver, {
      root: null,
      rootMargin: '0px',
      threshold: 1.0,
    });
    this.observer.observe(this.props.sentinelRef.current);
  }

  render() {
    return this.props.children;
  }
}

export default InfiniteScroll;
