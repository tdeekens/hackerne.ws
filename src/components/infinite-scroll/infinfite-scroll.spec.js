import React from 'react';
import { shallow } from 'enzyme';
import { InfiniteScroll } from './infinite-scroll';

const createTestProps = customProps => ({
  isLoading: false,
  onFetchMore: jest.fn(),
  sentinelRef: React.createRef(),
  children: <div />,
  ...customProps,
});

beforeAll(() => {
  global.IntersectionObserver = function() {
    return {
      observe: jest.fn(),
    };
  };
});

describe('rendering', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<InfiniteScroll {...props} />);
  });

  it('should render the passed children', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe('lifecycles', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<InfiniteScroll {...props} />);
  });

  describe('componentDidMount', () => {
    beforeEach(() => {
      wrapper.instance().componentDidMount();
      jest.spyOn(wrapper.instance().observer, 'observe');
      wrapper.instance().componentDidMount();
    });

    it('should observe the passed sentinel ref', () => {
      expect(wrapper.instance().observer.observe).toHaveBeenCalledWith(
        props.sentinelRef.current
      );
    });
  });
});

describe('interacting', () => {
  describe('handleObserver', () => {
    const createEntries = () => [
      {
        boundingClientRect: {
          y: 100,
        },
      },
    ];
    let props;
    let wrapper;

    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<InfiniteScroll {...props} />);
    });

    describe('given previous y is bigger then next y', () => {
      describe('when not loading', () => {
        beforeEach(() => {
          wrapper.setState({ prevY: 150 });
          wrapper.instance().handleObserver(createEntries());
        });
        it('should invoke `onFetchMore`', () => {
          expect(props.onFetchMore).toHaveBeenCalled();
        });
      });

      describe('when loading', () => {
        beforeEach(() => {
          props = createTestProps({
            isLoading: true,
          });
          wrapper = shallow(<InfiniteScroll {...props} />);
          wrapper.setState({ prevY: 150 });
          wrapper.instance().handleObserver(createEntries());
        });
        it('should not invoke `onFetchMore`', () => {
          expect(props.onFetchMore).not.toHaveBeenCalled();
        });
      });
    });

    describe('given previous y is smaller then next y', () => {
      describe('when not loading', () => {
        beforeEach(() => {
          wrapper.setState({ prevY: 50 });
          wrapper.instance().handleObserver(createEntries());
        });
        it('should not invoke `onFetchMore`', () => {
          expect(props.onFetchMore).not.toHaveBeenCalled();
        });
      });
    });
  });
});
