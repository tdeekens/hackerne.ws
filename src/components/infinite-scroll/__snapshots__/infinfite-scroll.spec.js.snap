// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should render the passed children 1`] = `<div />`;
